src/midi/midi_mt32.cpp

```diff
@@ -0,0 +1,516 @@
+
+	bool_prop = sec_prop.Add_bool("niceampramp", when_idle, true);
+	bool_prop->Set_help(
+	        "Toggles \"Nice Amp Ramp\" mode that improves amplitude ramp for sustaining instruments.\n"
+	        "Quick changes of volume or expression on a MIDI channel may result in amp jumps on real hardware.\n"
+	        "When \"Nice Amp Ramp\" mode is enabled, amp changes gradually instead.\n"
+	        "Otherwise, the emulation accuracy is preserved.\n"
+	        "Default is true.");
+}
+
+static mt32emu_report_handler_i get_report_handler_interface()
+{
+	class ReportHandler {
+	public:
+		static mt32emu_report_handler_version getReportHandlerVersionID(mt32emu_report_handler_i)
+		{
+			return MT32EMU_REPORT_HANDLER_VERSION_0;
+		}
+
+		static void printDebug(void *instance_data, const char *fmt, va_list list)
+		{
+			MidiHandler_mt32 &midiHandler_mt32 = *(
+			        MidiHandler_mt32 *)instance_data;
+			if (midiHandler_mt32.noise) {
+				char s[1024];
+				safe_sprintf(s, fmt, list);
+				LOG_MSG("MT32: %s", s);
+			}
+		}
+
+		static void onErrorControlROM(void *)
+		{
+			LOG_MSG("MT32: Couldn't open Control ROM file");
+		}
+
+		static void onErrorPCMROM(void *)
+		{
+			LOG_MSG("MT32: Couldn't open PCM ROM file");
+		}
+
+		static void showLCDMessage(void *, const char *message)
+		{
+			LOG_MSG("MT32: LCD-Message: %s", message);
+		}
+	};
+
+	static const mt32emu_report_handler_i_v0 REPORT_HANDLER_V0_IMPL = {
+	        ReportHandler::getReportHandlerVersionID,
+	        ReportHandler::printDebug,
+	        ReportHandler::onErrorControlROM,
+	        ReportHandler::onErrorPCMROM,
+	        ReportHandler::showLCDMessage,
+	        nullptr, // explicit empty function pointers
+	        nullptr,
+	        nullptr,
+	        nullptr,
+	        nullptr,
+	        nullptr,
+	        nullptr,
+	        nullptr,
+	        nullptr,
+	        nullptr};
+
+	static const mt32emu_report_handler_i REPORT_HANDLER_I = {
+	        &REPORT_HANDLER_V0_IMPL};
+
+	return REPORT_HANDLER_I;
+}
+
+static void make_rom_path(char pathName[],
+                          const char romDir[],
+                          const char fileName[],
+                          bool addPathSeparator)
+{
+	strcpy(pathName, romDir);
+	if (addPathSeparator)
+		strcat(pathName, "/");
+	strcat(pathName, fileName);
+}
+
+bool MidiHandler_mt32::Open(const char * /* conf */)
+{
+	service = new MT32Emu::Service();
+	uint32_t version = service->getLibraryVersionInt();
+	if (version < 0x020100) {
+		delete service;
+		service = nullptr;
+		LOG_MSG("MT32: libmt32emu version is too old: %s",
+		        service->getLibraryVersionString());
+		return false;
+	}
+	service->createContext(get_report_handler_interface(), this);
+	mt32emu_return_code rc;
+
+	Section_prop *section = static_cast<Section_prop *>(
+	        control->GetSection("mt32"));
+	const char *romDir = section->Get_string("romdir");
```

Comment:
Let's add TODO here to use std::string for romDir (TODO is good enough for now IMHO, we don't need to fix it right away).